### PR TITLE
test(unit): add test convention

### PR DESCRIPTION
**Pull request by thxCode**
**Fixes:**

#45

**Problem:**

Harvester needs some testings to guarantee its quality.

**Solution:**

Gives a unit testing convention and some test cases.

> Adding all cases is not the goal in this PR.

**Test plan:**

- `make test`

### CODE REVIEW

**Comment by orangedeng** [comment]
Need rebase

**Comment by thxCode** [comment]
@orangedeng rebased

File: Dockerfile.dapper [hunk 1]
```diff
@@ -1,29 +1,42 @@
...
+        gnupg \
+    && rm -rf /var/lib/apt/lists/*; \
+    \
+    curl -fsSL https://download.docker.com/linux/debian/gpg | apt-key add - >/dev/null; \
+    echo "deb [arch=$(dpkg --print-architecture)] https://download.docker.com/linux/debian buster stable" > /etc/apt/sources.list.d/docker.list; \
+    \
+    apt-get update -qq && apt-get install -y --no-install-recommends \
+        docker-ce-cli=5:19.03.* \
+    && rm -rf /var/lib/apt/lists/*; \
+    docker --version
```
Comment (orangedeng) [inline Dockerfile.dapper]: Version info will be mixed with docker build output. It is not easy for user to see this info. Maybe remove it or move it to ci script

File: Dockerfile.dapper [hunk 2]
```diff
@@ -1,29 +1,42 @@
...
+## install golangci
+RUN if [ "$(go env GOARCH)" = "amd64" ]; then \
+        curl -sL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b "$(go env GOPATH)/bin" v1.27.0; \
+        golangci-lint --version; \
```
Comment (orangedeng) [inline Dockerfile.dapper]: Same as docker --version.

File: scripts/test [hunk 3]
```diff
@@ -1,7 +1,18 @@
 #!/bin/bash
 set -e
 
-cd $(dirname $0)/..
+# The root of the harvester directory
+ROOT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd -P)"
 
-echo Running tests
-go test -cover -tags=test ./...
+mkdir -p "${ROOT_DIR}/bin"
```
Comment (orangedeng) [inline scripts/test]: Do we need `bin` here?

Comment (thxCode) [reply scripts/test]: No, removed.

File: Dockerfile.dapper (hunk 2 above)
Comment (thxCode) [reply Dockerfile.dapper]: Updated

File: Dockerfile.dapper (hunk 1 above)
Comment (thxCode) [reply Dockerfile.dapper]: Updated

**Review by orangedeng — APPROVED** [review]
LGTM

File: pkg/controller/master/keypair/controller_test.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+			}
+		}
+
+		assert.Equal(t, tc.expected, actual, "case %q", tc.name)
+	}
+}
+
+func generateSSHPublicKey() (pk string, fingerprint string) {
+	var key, _ = rsa.GenerateKey(rand.Reader, 2048)
+	var pubKey, _ = ssh.NewPublicKey(&key.PublicKey)
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/keypair/controller_test.go]: Updated

File: pkg/controller/master/keypair/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+				},
+				err: nil,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		var clientset = fake.NewSimpleClientset()
+		if tc.given.keyPair != nil {
+			_ = clientset.Tracker().Add(tc.given.keyPair)
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/keypair/controller_test.go]: Leave a message for others: here returns an error.
I assert this now.

File: pkg/controller/master/keypair/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/crypto/ssh"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+
+	vmapis "github.com/rancher/harvester/pkg/apis/harvester.cattle.io/v1alpha1"
+	"github.com/rancher/harvester/pkg/generated/clientset/versioned/fake"
+	vmtype "github.com/rancher/harvester/pkg/generated/clientset/versioned/typed/harvester.cattle.io/v1alpha1"
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/controller/master/keypair/controller_test.go]: Updated

File: Dockerfile.dapper [hunk 7]
```diff
@@ -1,29 +1,40 @@
...
-RUN mkdir -p /go/src/golang.org/x && \
-    cd /go/src/golang.org/x && git clone https://github.com/golang/tools && \
-    git -C /go/src/golang.org/x/tools checkout -b current aa82965741a9fecd12b026fbb3d3c6ed3231b8f8 && \
-    go install golang.org/x/tools/cmd/goimports
-RUN rm -rf /go/src /go/pkg
-RUN if [ "${ARCH}" == "amd64" ]; then \
-        curl -sL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s v1.15.0; \
+# -- for make rules
+## install docker client
+RUN apt-get update -qq && apt-get install -y --no-install-recommends \
```
[the remark this answers is not in the document]
Comment (thxCode) [reply Dockerfile.dapper:10]: Follow the domain rule: use a `RUN` to install one tool, and then describe its dependency installation in the same `RUN`.

**Review by gitlawr — APPROVED** [review]

**Pull request merged**